add deepcrawl API

mayorbyrne opened the pull request:
- https://level7.atlassian.net/browse/PLAT-1787

Commit:
add deepcrawl API

bmoeskau commented on README.md:
```diff
@@ -1,2 +1,17 @@
+
+var deepCrawl = new DeepCrawl({
+  apiKey: '[your api key]',
+  apiId: '[your api id]',
```
You'll probably want to default these to undefined so they'll throw in the constructor by default

mayorbyrne replied:
they should.  see line 10 of deepcrawl.js

bmoeskau replied:
Oh, I just realized this is the README. I thought this was a config file for some reason

Commit:
rename .es > .js

bmoeskau commented on lib/deepcrawl.es:
```diff
@@ -0,0 +1,120 @@
+    this.apiId = cfg.apiId;
+    this.getSessionToken = cfg.getSessionToken;
+    this.accountId = cfg.accountId;
+    this.baseUrl = 'https://prod-1-dc-api-oopeix3r.deepcrawl.com';
```
This should come from config too

bmoeskau commented on lib/schema.js:
```diff
@@ -0,0 +1,68 @@
+'use strict';
+module.exports = {
+  projects: {
+    url: '/accounts/{accountId}/projects',
```
`route` or `path` might make more sense than `url` here

bmoeskau commented on lib/schema.js:
```diff
@@ -0,0 +1,68 @@
+      requiredFields: ['name', 'sitePrimary']
+    },
+    read: {
+      url: '/{projectId}',
```
By the same token, this is also not a `url`. Maybe `path` makes the most sense.

However in traditional REST terms, this is not technically needed at this level. As a standard all GET, PUT, PATCH and DELETE operations expect an id as the last part of the path, so you can probably omit it here and default to this structure in the code. You should only have to include an explicit path for these operations if they deviate from standard REST format in some cases.

bmoeskau replied:
Unless this is useful for mapping `projectId` specifically I guess, for easily mapping that name from the data as the "id" value. It's your call, but if the ids can be mapped generically that would save boilerplate schema code.

bmoeskau commented on lib/schema.js:
```diff
@@ -0,0 +1,68 @@
+    requiredFields: ['accountId', 'projectId'],
+    list: {
+      type: 'GET'
+    },
```
Another nice shortcut to consider would be checking these first to see if they are string, then defaulting the value to being the verb with defaults in that case. If non-string then go through the existing logic. It's minor, but would allow simpler schema code like `list: 'GET'` here

bmoeskau commented on lib/schema.js:
```diff
@@ -0,0 +1,68 @@
+    delete: {
+      url: '/{projectId}',
+      type: 'DELETE',
+      requiredFields: ['projectId']
```
Also for GET, PUT, PATCH, DELETE the id value will always be required. Really this could be as simple as `delete: 'DELETE'` with the path and required field defaulted programmatically.

mayorbyrne replied:
GET and POST /projects without the trailing /{projectId} does not require the projectId, that's why i had them separated a bit.

bmoeskau replied:
Oh that's true, GET does lists too. Yeah cool, but for PUT, PATCH, DELETE I believe they are always required.

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,120 @@
+    this.apiId = cfg.apiId;
+    this.getSessionToken = cfg.getSessionToken;
+    this.accountId = cfg.accountId;
+    this.baseUrl = 'https://prod-1-dc-api-oopeix3r.deepcrawl.com';
```
This should come from config too

Commit:
address comments
- change url to "route"
- add simple unit test
- simplify schema
- add "id" field to endpoints (read, update, delete)
- baseUrl should come from config

Commit:
tweak unit test

Commit:
simplify schema
- try and fix unit test

Commit:
add circle.yml

bmoeskau commented on lib/schema.js:
```diff
@@ -0,0 +1,33 @@
+    id: 'userAgentCode',
+    list: 'GET',
+    read: 'GET'
+  }
```
This is looking pretty good. A few suggestions:
- Needs a more descriptive filename. Maybe `api_schema_v2.0.0.js`?
- I would add a top-level property on the schema itself like `version: '2.0.0'` so it can be queried programmatically
- I would move projects, etc. under a top-level property called "resources" (REST terminology for the main domain objects in the API)
- Under each resource I would move the actions (list, create, etc.) into a containing property called "actions"

The containing properties add a tiny amount more nesting, but make it much clearer what is what when scanning this file, plus make it easier to iterate over the specific things you intend to without having to delete things like you're doing currently (a little weird)

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+      // remove the endpoint route, required fields, and id from the schema
+      delete schema[endpoint].requiredFields;
+      delete schema[endpoint].route;
+      delete schema[endpoint].id;
```
Shouldn't need to delete these. Per my comments below you should be able to just do
`for (const action in schema[endpoint].actions)`

bmoeskau replied:
As a side note, you're currently mixing the meaning of "method" between HTTP request methods (GET, PUT, etc) and programmatic API methods (list, create, etc) which confuses things a little. I think by changing the API terminology to "actions" and keeping the http terminology as "methods" it should keep things clearer.

Commit:
address comments
- add "resources" and "actions" to schema
- remove required fields apiKey and apiId since we don't use them currently
- move schemas to separate folder
- fix unit tests

Commit:
return res body if success
- handle 409 / 422 errors

bmoeskau commented on README.md:
```diff
@@ -1,2 +1,27 @@
 # deepcrawl
 A DeepCrawl API wrapper for Node
+
+This module currently assumes you are refreshing and storing your deepCrawl session token on a regular basis (via your provided api id and api key), and have a means of retrieving it.
```
"DeepCrawl"

bmoeskau commented on schemas/v2.0.0.js:
```diff
@@ -0,0 +1,43 @@
+'use strict';
+module.exports = {
+  version: '2.0.0',
+  baseUrl: 'https://prod-1-dc-api-oopeix3r.deepcrawl.com',
```
`baseUrl` is back! Still don't think it belongs here since this schema is not **environment** specific. This url should always be passed from an env-specific config

mayorbyrne replied:
don't you think this is going to be the same for everyone once they actually figure out the api?

bmoeskau replied:
It won't even be the same for **us** since we know it's going to change before this goes live. When you already know something will be variable like that, it belongs in config. Also consider that different versions of the schema (2.0.1, 2.1.0, etc.) will almost certainly use the same base url, and so you wouldn't want that duplicated. You'd want it configured at the top level and passed into the API code.

bmoeskau replied:
The schema only specifies the syntax for the API, not where that API lives. It can move from one base url to another, but as long as the version stays the same the schema should not require any changes.

bmoeskau commented on README.md:
```diff
@@ -1,2 +1,27 @@
+  accountId: '[your account id]' // must be passed here or with each respective call
+});
+
+var API = deepCrawl.getAPI();
```
I keep going back and forth on whether I like this `getAPI` method (that I suggested). I'm starting to lean back toward 'no'. What does everyone else think?

mayorbyrne replied:
the other option being return the API from the constructor?

bmoeskau replied:
yes

mayorbyrne replied:
Although I like the idea of just making one call to get the API, I do think (as you mentioned previously) the getAPI method gives us more flexibility moving forward, in case we ever need to store anything top level, especially if we ever added on-the-fly sessionToken refreshing / storage.

bmoeskau commented on index.js:
```diff
@@ -0,0 +1,2 @@
+'use strict';
+module.exports = require('./lib/deepcrawl');
```
At this point we might as well nuke the separate file and just put the actual code here. Any reason not to?

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+const BPromise = require('bluebird'),
+  schema = require('../schemas/v2.0.0'),
+  needle = BPromise.promisifyAll(require('needle')),
+  humps = require('humps'),
```
I can't read this line without laughing

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+    this.apiId = cfg.apiId;
+    this.apiKey = cfg.apiKey;
+    this.getSessionToken = cfg.getSessionToken;
+    this.accountId = cfg.accountId;
```
Should we be throwing here if any of these are null?

mayorbyrne replied:
i'm not throwing, because none of them are technically required at this point.  i check for sessionToken on each method call, if it doesn't exist and getSessionToken also doesn't exist, then i throw an error.  same with accountId.  they can pass that in to each call as well.  apiId and apiKey aren't required right now.  they honestly don't have to be here.

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+    }
+    if (res.statusCode === 200 || res.statusCode === 201 || res.statusCode === 202 || res.statusCode === 204) {
+      return res.body;
+    }
```
else... ?

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+    if (res.statusCode === 422) {
+      throw new Error(`There was a validation error. ${JSON.stringify(res.body)}`);
+    }
+    if (res.statusCode === 200 || res.statusCode === 201 || res.statusCode === 202 || res.statusCode === 204) {
```
I assume this could just be `if (String(res.statusCode).startsWith('2')) {`

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+    this.accountId = opts.accountId || this.accountId;
+    // create the base API
+    const API = this.createEndpoints(schema);
+    this.addMethodsToEndpoints(API, schema);
```
At the moment the `accountId` is cached and the API is dynamically created on each call to this method. Ultimately we'll probably want to reverse this -- cache the API after initial creation, and only create a new instance if a different version is requested (future). For now it should never change, and the only things that could change from one call to the next are the variables like account id, API key, etc. The basic schema should never change during the lifetime of this class instance.

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+      API[resource] = {};
+    }
+    return API;
+  }
```
I would nuke this method, I don't think it's needed. See comments below in `addMethodsToEndpoints`

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  addMethodsToEndpoints (API, schema) {
+    for (const resource in API) {
```
I think this method should be renamed to something more general like `parseSchema` which is really all it's doing. This initial loop can simply be `for (const resource in API.resources) {` to avoid the extraneous `createEndpoints` method -- that's the whole beauty of adding that `resources` level so you no longer have to explicitly parse them out -- they are already there ;)

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+
+  addMethodsToEndpoints (API, schema) {
+    for (const resource in API) {
+      const resourceUrl = this.baseUrl + schema.resources[resource].route,
```
I would extract this line into a separate (testable) method, and also add a wrapping url parsing function around it in case someone specifies an invalid url format in the schema, leaves off a leading slash accidentally, etc. You should have some kind of proper handling anytime you are building urls from variables.

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+        let actionRequiredFields = requiredFields,
+          actionUrl = resourceUrl;
+        // check to see whether the resource action has extra data associated with it
+        const isString = typeof schema.resources[resource].actions[action] === 'string';
```
I would rename the loop const to `actionName` and then immediately create an inner `const action = schema.resources[resource].actions[actionName]` to represent the action string/object that you can use throughout this block.

bmoeskau replied:
In fact, instead of this `isString` var, a more elegant approach (to avoid the conditional checking below) is to convert the string into a defaulted object, e.g.:

```
  if (typeof action === 'string') {
    action = {
      type: action,
      requiredFields: []
    };
  }
```

It's slightly more code here, but then less (and much simpler) code later.

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+        // we will append the id field to the required fields for the action, and
+        // add /{id} to the url, which will be replaced later with the passed in
+        // field
+        if (action === 'read' || action === 'update' || action === 'delete') {
```
How about `if (requiredFields.length && action.match(/read|update|delete/i)) {`?

bmoeskau replied:
Well I guess you still want to append the id even if requiredFields is empty, but the regex is still a nicer check.

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+    return (options) => {
+      options = options || {};
+      if (!options.sessionToken && !this.getSessionToken) {
+        throw new Error(`"sessionToken" or "getSessionToken" method required`);
```
I would think we'd throw this right up front rather than waiting until we're knee-deep into the schema parsing.

mayorbyrne replied:
We can check whether they passed in a getSessionToken method to the ctor, but not whether they call each API method with a sessionToken or not.  That's the reason I'm making the check here, to see if they passed neither.  Unless you want to make the getSessionToken method required in the ctor, I don't see how else to accomplish this.

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  generateMethod (requiredFields, url, method) {
+    return (options) => {
```
Can you add some inline comments here that returning this function as a closure is the key to constructing it as a dynamic method call so that future knuckleheads like me won't also be confused.

bmoeskau commented on lib/deepcrawl.js:
```diff
@@ -0,0 +1,137 @@
+    return API;
+  }
+
+  generateRequest (url, options, method) {
```
I would rename this to "doRequest" to make clear it's actually calling the API, not returning a dynamic "request" object or closure for later use (which is what this sounds like)

Commit:
address comments
- make baseUrl a config value, validate in ctor
- generateRequest > performRequest
- validate resourceUrl
- add parseSchema method, greatly simplify innards
- handle all responses
- change "type" to "method" in schema

Commit:
nuke /lib dir
- move all logic into index.js

bmoeskau commented on index.js:
```diff
@@ -0,0 +1,211 @@
+      throw new Error('No baseUrl provided');
+    }
+    if (url[url.length - 1] === '/') {
+      throw new Error(`${url} should not end in a slash.`);
```
Don't throw here, just trim it off if needed

bmoeskau commented on index.js:
```diff
@@ -0,0 +1,211 @@
+      throw new Error(`The route ${route} incorrectly ends with a trailing slash. Routes should only begin with a slash.`);
+    }
+    if (route[0] !== '/') {
+      throw new Error(`The route ${route} incorrectly starts without a slash.  Routes should begin with a slash.`);
```
Again, be nice here and just fix it if you can, rather than throwing

bmoeskau commented on index.js:
```diff
@@ -0,0 +1,211 @@
+          action.requiredFields = [id].concat(action.requiredFields);
+          action.url = resourceUrl + `/{${id}}`;
+        }
+        API[resource][actionName] = this.generateMethod(action);    
```
This method is finally looking pretty solid. The logic reads very cleanly.

bmoeskau commented:
This is looking good. Just add a few tests that prove that the schema parsing is working down to the resource/action level and let's get it merged.

Commit:
address comments
- be nice and fix the routes/baseUrl rather than throw error
- add unit tests for schema parsing

Commit:
grunt

Commit:
fix typos

Commit:
add the rest of the API endpoints to the schema
- querystring GET options for endpoints like /domain_responses

Commit:
comment out querystring options for now

Commit:
fix url for GET /domainResponses

bmoeskau committed:
Minor README update

Commit:
extend report template with name field

Commit:
add gruntfile
- grunt; beautify

Commit:
remove unnecessary

toddbluhm committed:
Cleanup for eslint/jsbeautify and settings tweaks

bmoeskau committed:
Tweak schema loading

- Remove redundant "v" prefix from schema file
- Simplify the API by requiring a version number, rather than a fully-loaded schema object, by default (parseSchem() supports both)
- Make the apiVersion config explicitly required -- we never want to implicitly load a version that the consumer is not expecting

bmoeskau committed:
Nuke dc.getAPI()

Tthe class now sets an `api` property on creation, or when calling loadSchema(), so no need for an extra method call each time. The `api` property is needed to avoid potential future name collisions between the DC API and our class API.

bmoeskau committed:
Move version to the .api level

Makes more sense there, and is still easily accessible as `dc.api.version`

bmoeskau committed:
Add custom schema loading test, remove extraneous describe() level

bmoeskau committed:
Change "needle" prop to "requestLib"

Should never use a specific library name as a property like that. Needle could be replaced with any lib that supports the basic get/post/patch/delete verbs (e.g. request) and it should not affect the DC class API. Instead default it to needle internally but name it generically on the class.

Also fixed the indentation on the "crawls" level of unit tests

bmoeskau committed:
Replace getSessionToken with sessionToken

bmoeskau committed:
Add getSessionToken()

bmoeskau committed:
Clarify API beta status
